src/Microsoft.DotNet.Cli.Utils/CommandResolution/DotnetToolsCommandResolver.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Collections.Generic;
+using Microsoft.DotNet.PlatformAbstractions;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    public class DotnetToolsCommandResolver : ICommandResolver
+    {
+        private string _dotnetToolPath;
+
+        public DotnetToolsCommandResolver(string dotnetToolPath = null)
+        {
+            if (dotnetToolPath == null)
+            {
+                _dotnetToolPath = Path.Combine(ApplicationEnvironment.ApplicationBasePath,
+                    "DotnetTools");
+            }
+            else
+            {
+                _dotnetToolPath = dotnetToolPath;
+            }
+        }
+
+        public CommandSpec Resolve(CommandResolverArguments arguments)
+        {
+            if (string.IsNullOrEmpty(arguments.CommandName))
+            {
+                return null;
+            }
+
+            var packageId = new DirectoryInfo(Path.Combine(_dotnetToolPath, arguments.CommandName));
+            if (!packageId.Exists)
+            {
+                return null;
+            }
+
+            var version = packageId.GetDirectories()[0];
+            var dll = version.GetDirectories("tools")[0]
```

Comment:
with assumption of packageid=packagename, there is only 1 tfm, only 1 rid

Reply:
We should document this in the CLI repo somewhere. Making it clear that this is a requirement for native verbs in the CLI implemented through global commands packages.

Reply:
I added a set of test to enforce the package structure. And the test name is pretty much the contract. If will break during a bad insert, the test name should indicate the problem very well.

https://github.com/dotnet/cli/pull/8606/files#diff-5a5e8d0e463a1b61fe06d23fab49eca5

Do we still need a document? If so where should I put it